uniffi_macros/src/error.rs

```diff
@@ -0,0 +1,170 @@
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    punctuated::Punctuated,
+    Data, DeriveInput, Index, Token, Variant,
+};
+use uniffi_meta::{ErrorMetadata, VariantMetadata};
+
+use crate::{
+    enum_::{enum_ffi_converter_impl, variant_metadata},
+    util::{
+        assert_type_eq, chain, create_metadata_static_var, either_attribute_arg, AttributeSliceExt,
+        UniffiAttribute,
+    },
+};
+
+pub fn expand_error(input: DeriveInput, module_path: Vec<String>) -> TokenStream {
+    let variants = match input.data {
+        Data::Enum(e) => Ok(e.variants),
+        _ => Err(syn::Error::new(
+            Span::call_site(),
+            "This derive currently only supports enums",
+        )),
+    };
+
+    let ident = &input.ident;
+    let attr = input.attrs.parse_uniffi_attributes::<ErrorAttr>();
+    let ffi_converter_impl = match &attr {
+        Ok(a) if a.flat.is_some() => flat_error_ffi_converter_impl(variants.as_ref().ok(), ident),
+        _ => enum_ffi_converter_impl(variants.as_ref().ok(), ident),
+    };
+
+    let meta_static_var = match (&variants, &attr) {
+        (Ok(vs), Ok(a)) => Some(match error_metadata(ident, vs, module_path, a) {
+            Ok(metadata) => create_metadata_static_var(ident, metadata.into()),
+            Err(e) => e.into_compile_error(),
+        }),
+        _ => None,
+    };
+
+    let type_assertion = assert_type_eq(ident, quote! { crate::uniffi_types::#ident });
+    let variant_errors: TokenStream = match variants {
+        Ok(vs) => vs
+            .iter()
+            .flat_map(|variant| {
+                chain(
+                    variant.attrs.attributes_not_allowed_here(),
+                    variant
+                        .fields
+                        .iter()
+                        .flat_map(|field| field.attrs.attributes_not_allowed_here()),
+                )
+            })
+            .map(syn::Error::into_compile_error)
+            .collect(),
+        Err(e) => e.into_compile_error(),
+    };
+    let attr_error = attr.err().map(syn::Error::into_compile_error);
+
+    quote! {
+        #ffi_converter_impl
+
+        #[automatically_derived]
+        impl ::uniffi::FfiError for #ident {}
```

Comment:
I have been trying to understand why we have this trait.
Is this just a marker?
It doesn't seem to be used anywhere or am I missing something?